### ISSUE TITLE
Add DELTREE command

### ISSUE THREAD

**Issue by NicknineTheEagle** [dosbox-staging/dosbox-staging#626]
**Is your feature request related to a problem? Please describe.**
It is currently not possible to delete a non-empty directory from within DosBox command line.

**Describe the solution you'd like**
Add DELTREE command from DOS. It deletes a directory with all the files and sub-directories within it: https://www.computerhope.com/deltree.htm

**Describe alternatives you've considered**
This can be worked around by just deleting a directory from host OS or using a DOS file manager like Norton Commander but switching back and forth every time I want to delete a directory is inconvenient.

**Additional context**
Starting from Windows 2000, DELTREE was merged into RMDIR command as /S switch. Not sure if we want to add that or not.

**Comment by kcgen** [comment]
I only have once concern, and that's minimizing accidental blast radius. For example, some users mount their entire DOS collection as C:\; some go further and mount all of `C:\Program Files (x86)` as C:\; and (I know in the past) some have mounted their either root directory as C:\. 

 would hate for someone to accidentally run DELTREE and lose their entire collection and/or ripple into removing host programs and settings.

Although I haven't come across them, do some games or their installers use `DELTREE`?

By not providing `DELTREE`, it forces users to blow away recursive trees using their host OS (GUI or command-line), in which case we're not responsible for the damage.  On the flip side, the user will gain whatever safety nets are in place such as warning pop-ups, intermediate trash can location, and so on.

**Comment by NicknineTheEagle** [comment]
Yikes, I did not think of that. Still, we should not gut the functionality because of some users' stupidity. Maybe we could put some kind of additional safeguard in? Like a mount parameter or environment variable?

**Comment by NicknineTheEagle** [comment]
> Although I haven't come across them, do some games or their installers use DELTREE?

Not to my knowledge.

**Comment by kcgen** [comment]
> Still, we should not gut the functionality because of some users' stupidity. 

I agree and I like your suggestion of a safe-guard. Power-users might scoff at any protection, but it's not like users will be running this command repeatedly.  A double-check (with red or yellow color), perhaps with a nice summary of the quantity and total MB (or even GB) about to be blown away could be presented.  

At that point.. I would feel we've done our due-diligence, and the rest is in the hands of the user.

**Comment by kcgen** [comment]
We would also have to detect and prevent traversing logical symlinks to other directories.  For example, on non-Windows platform, a directory can be a symlink to another drive's root partition.. meaning if we didn't handle this properly, we could blow away other drive (or directory) contents as well.

**Comment by MasterO2** [comment]
> 
> 
> I have one concern, and that's the size of the blast radius. For example, some users mount their entire DOS collection as `C:\`; some go further and mount all of `C:\Program Files (x86)` as `C:\`; and (I know in the past) some have mounted their either root directory as `C:\`.
> 
> would hate for someone to lose their entire collection and/or host programs and settings.
> 
> Although I haven't come across them, do some games or their installers use `DELTREE`?
> 
> By not providing `DELTREE`, it forces users to blow away recursive trees using their host OS (GUI or command-line), in which case we're not responsible for the damage. Also by using their host mechanism, they gain whatever safety nets are in place such as warning pop-ups, intermediate trash can location, and so on (all of which would be tedious or even impossible for us to activate inside of dosbox)

@kcgen Let them use DELTREE, but dosbox-staging itself should not allow the end user to mount the following directories within dosbox-staging as dosbox-staging's C drive:

c:\program files\
C:\WINDOWS\ or any of its subfolders

So that the end user does not accidentally screw up their own actual system.

**Comment by NicknineTheEagle** [comment]
DELTREE already requests confirmation by default (unless you pass /Y switch), we can confirmation message bigger and badder.

**Comment by Wengier** [comment]
The confirmation message displayed by MS-DOS's DELTREE command is "Delete directory <name> and all its subdirectories [yn]?", and its help message says "Note: Use DELTREE cautiously. Every file and subdirectory within the specified directory will be deleted.". When I implemented the DELTREE command earlier (https://github.com/joncampbell123/dosbox-x/commit/03c97ebc902db0b049bfebd81319072a450bc076) I tried to use the same message as in MS-DOS.

**Comment by kcgen** [comment]
@dreamer ; how close are we to C++17? (u16.04 goes EOL in about 6.5 months). 

`std::filesystem::remove_all(..)` could be a very elegant way to implement this.

https://en.cppreference.com/w/cpp/filesystem/remove

**Comment by dreamer** [comment]
C++17 definitely won't happen for 0.76.0; perhaps for 0.77.0.

I need to re-evaluate what's available in Steam Runtime environment (last time I checked it was still limited to GCC 5.4, which had no C++17 support, thus libraries provided by the runtime would be ABI-incompatible with dosbox-staging. I don't know how the situation looks like on AmigaOS nor Haiku (we need to at least check before moving to new std). FreeBSD probably has C++17 support (but I need to double-check as well).

**Comment by Wengier** [comment]
Indeed, there is never a standardized and cross-platform way of deleting the whole directory in C/C++ before C++17. So I and other developers have to do the actual file deletions manually. For Windows you can just use SHFileOperation for deleting all files but it is not cross-platform in any way.

**Comment by grapeli** [comment]
> So that the end user does not accidentally screw up their own actual system. Add root (/) in Linux to that list as well.

@MasterO2 You can. Try under [JSLinux](https://bellard.org/jslinux/vm.html?cpu=riscv64&url=buildroot-riscv64-xwin.cfg&graphic=1&mem=256). Load this file [dosbox-staging.riscv64.zip](https://github.com/dosbox-staging/dosbox-staging/files/5289092/dosbox-staging.riscv64.zip), open a terminal unpack, run and check.
```
unzip dosbox-staging.riscv64.zip
./dosbox
```

**Comment by MasterO2** [comment]
@grapeli 

This?

![This](https://i.ibb.co/9hPXRk0/that.png)

**Comment by grapeli** [comment]
Yes. Of course, you won't remove the symbolic links themselves, the special filesystem 'proc', 'sys' or 'dev'. Use deltree from freedos or [rm -rf](ftp://ftp.delorie.com/pub/djgpp/current/v2gnu/shl2011br3.zip).

**Comment by dreamer** [comment]
`/proc` is additionally protected from editing from within DOSBox - that was the security issue, that triggered DOSBox 0.74-3 release in 2019; the fix is available in all versions of dosbox-staging. Anything potentially risky in `/sys` and `/dev` is already protected by requiring root access.

On Windows, AFAIK, `C:\Windows\` cannot be modified by userspace applications  without administrator access since Windows XP (or maybe Vista?).

Small reminder: if anyone finds a security issue within dosbox-staging, then [here's our security.txt file](https://dosbox-staging.github.io/security.txt) with info on how to report security-related issues without disclosing them publically.

**Comment by Burrito78** [comment]
I have come across one real-world example where a missing DOS program prevents the correct installation of a game: Tie Fighter floppy version needs debug.

https://github.com/dosbox-staging/dosbox-staging/issues/993#issuecomment-830352867

In that case i'd suggest we either need:

1. A little debug program shipped inside Dosbox-Staging that gets the job done regarding Tie Fighter.
2. A debug stub that outputs "This program needs debug to work properly! Get it here: ..." and pause the screen. Could be that this message isn't seen however because it would be executed in the background without output maybe. Would need to test that.

**Comment by kcgen** [comment]
Thank you @Burrito78,
This is in-scope and your suggestions are excellent; new ticket opened.
